docs/section-8-code-navigation-systems.md

```diff
@@ -0,0 +1,101 @@
+---
+title: Code Navigation Systems
+permalink: code-navigation-systems
+---
+
+# Code Navigation Systems
+
+Tree-sitter can be used in conjunction with its [tree query language](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries) as a part of code navigation systems. An example of such a system can be seen in the `tree-sitter tag` command, which emits a textual dump of the interesting syntactic nodes in its file argument. A notable application of this is GitHub's support for [search-based code navigation](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github#precise-and-search-based-navigation). This document exists to describe how to integrate with such systems, and how to extend this functionality to any language with a Tree-sitter grammar.
+
+## Tagging and captures
+
+*Tagging* is the act of identifying the entities that can be named in a program. We use Tree-sitter queries to find those entities. Having found them, you use a syntax capture to label the entity and its name.
+
+The essence of a given tag lies in two pieces of data: the _kind_ of entity that is matched (usually a definition or a reference) and the _role_ of that entity, which describes how the entity is used (i.e. whether it's a class definition, function call, variable reference, and so on). Our convention is to use a syntax capture following the `@kind.role` capture name format, and another inner capture, always called `@name`, that pulls out the name of a given identifier.
```

Comment:
It looks like `kind` and `role` might be reversed in this part. Role is for definition vs reference, right?

Reply:
Doug’s original suggestion re. “kind” and “role” was ambiguous (or I, in foolish haste, misread it) but I think a case could be made for either. Though upon consideration I am gonna switch them around, yeah.